### Отчет по лабораторной работе №2

**Тема:** Поиск экстремума функции методами половинного деления, золотого сечения, хорд и Ньютона

#### Задача

Найти минимум функции $ f(x) = \ln(1 + x^2) - \sin(x) $ на интервале $[0, \frac{\pi}{4}]$ с точностью 0.001 (вручную) и 0.0001 (программно).


In [80]:
# Начальные определения
import math


def f(x):
    return math.log(1 + x**2) - math.sin(x)


def f_prime(x):
    return (2 * x) / (1 + x**2) - math.cos(x)


def f_double_prime(x):
    return (2 - 2 * x**2) / (1 + x**2) ** 2 + math.sin(x)


a, b = 0, math.pi / 4

precision = 4

epsilon = 10 ** (-precision)

## 1. Метод половинного деления (бисекции)

### Алгоритм:

1. Задаем интервал $[a, b]$, где $a = 0$, $b = \frac{\pi}{4} \approx 0.7854$
2. Проверяем, что $f'(a) \cdot f'(b) < 0$ (производная меняет знак на интервале)
3. Находим середину интервала $c = \frac{a + b}{2}$
4. Вычисляем $f'(c)$
5. Если $|f'(c)| < \varepsilon$ или $|b - a| < \varepsilon$, то $c$ - приближенное значение минимума
6. Иначе, если $f'(c) \cdot f'(a) > 0$, то $a = c$, иначе $b = c$
7. Повторяем шаги 3-6

Для функции $f(x) = \ln(1 + x^2) - \sin(x)$ на интервале $[0, \frac{\pi}{4}]$ с точностью $\varepsilon = 0.001$.

Первая производная: $f'(x) = \frac{2x}{1+x^2} - \cos(x)$

Для поиска минимума нам нужно найти точку, где $f'(x) = 0$.

### Проверка знаков производной на концах интервала:

- $f'(0) = \frac{2 \cdot 0}{1+0^2} - \cos(0) = 0 - 1 = -1 < 0$
- $f'(\frac{\pi}{4}) = f'(0.7854) = \frac{2 \cdot 0.7854}{1+0.7854^2} - \cos(0.7854) \approx \frac{1.5708}{1.6168} - 0.7071 \approx 0.9716 - 0.7071 \approx 0.2645 > 0$

Поскольку $f'(0) < 0$ и $f'(\frac{\pi}{4}) > 0$, производная меняет знак на интервале, и внутри него есть точка минимума.

### Вычисление метода половинного деления:

**Итерация 1:**

- $a_0 = 0$, $b_0 = 0.7854$
- $c_1 = \frac{a_0 + b_0}{2} = \frac{0 + 0.7854}{2} = 0.3927$
- $f'(c_1) = f'(0.3927) = \frac{2 \cdot 0.3927}{1+0.3927^2} - \cos(0.3927) \approx \frac{0.7854}{1.1542} - 0.9239 \approx 0.6804 - 0.9239 \approx -0.2435 < 0$
- Так как $f'(c_1) < 0$ (тот же знак, что и $f'(a_0)$), то $a_1 = c_1 = 0.3927$, $b_1 = b_0 = 0.7854$

**Итерация 2:**

- $a_1 = 0.3927$, $b_1 = 0.7854$
- $c_2 = \frac{a_1 + b_1}{2} = \frac{0.3927 + 0.7854}{2} = 0.5891$
- $f'(c_2) = f'(0.5891) \approx \frac{2 \cdot 0.5891}{1+0.5891^2} - \cos(0.5891) \approx \frac{1.1782}{1.3470} - 0.8263 \approx 0.8747 - 0.8263 \approx 0.0484 > 0$
- Так как $f'(c_2) > 0$ (тот же знак, что и $f'(b_1)$), то $a_2 = a_1 = 0.3927$, $b_2 = c_2 = 0.5891$

**Итерация 3:**

- $a_2 = 0.3927$, $b_2 = 0.5891$
- $c_3 = \frac{a_2 + b_2}{2} = \frac{0.3927 + 0.5891}{2} = 0.4909$
- $f'(c_3) = f'(0.4909) \approx \frac{2 \cdot 0.4909}{1+0.4909^2} - \cos(0.4909) \approx \frac{0.9818}{1.2410} - 0.8739 \approx 0.7911 - 0.8739 \approx -0.0828 < 0$
- Так как $f'(c_3) < 0$ (тот же знак, что и $f'(a_2)$), то $a_3 = c_3 = 0.4909$, $b_3 = b_2 = 0.5891$

**Итерация 4:**

- $a_3 = 0.4909$, $b_3 = 0.5891$
- $c_4 = \frac{a_3 + b_3}{2} = \frac{0.4909 + 0.5891}{2} = 0.5400$
- $f'(c_4) = f'(0.5400) \approx \frac{2 \cdot 0.5400}{1+0.5400^2} - \cos(0.5400) \approx \frac{1.0800}{1.2916} - 0.8526 \approx 0.8361 - 0.8526 \approx -0.0165 < 0$
- Так как $f'(c_4) < 0$ (тот же знак, что и $f'(a_3)$), то $a_4 = c_4 = 0.5400$, $b_4 = b_3 = 0.5891$

**Итерация 5:**

- $a_4 = 0.5400$, $b_4 = 0.5891$
- $c_5 = \frac{a_4 + b_4}{2} = \frac{0.5400 + 0.5891}{2} = 0.5646$
- $f'(c_5) = f'(0.5646) \approx \frac{2 \cdot 0.5646}{1+0.5646^2} - \cos(0.5646) \approx \frac{1.1292}{1.3188} - 0.8394 \approx 0.8563 - 0.8394 \approx 0.0169 > 0$
- Так как $f'(c_5) > 0$ (тот же знак, что и $f'(b_4)$), то $a_5 = a_4 = 0.5400$, $b_5 = c_5 = 0.5646$

После 5 итераций получаем интервал $[0.5400, 0.5646]$, в котором находится минимум. Приближенное значение минимума: $x \approx 0.5523$.

Длина итогового интервала: $0.5646 - 0.5400 = 0.0246 > 0.001$, поэтому для достижения заданной точности потребуются дополнительные итерации.


In [81]:
def bisection_extremum(a, b, epsilon, max_iterations=1000) -> float:
    """
    Находит экстремум функции методом половинного деления.

    Параметры:
    a, b -- границы интервала
    epsilon -- требуемая точность
    max_iterations -- максимальное число итераций

    Возвращает:
    x -- приближенное значение экстремума
    """
    if f_prime(a) * f_prime(b) >= 0:
        raise ValueError("Производная должна менять знак на интервале [a, b]")

    iteration = 0
    while (b - a) > epsilon and iteration < max_iterations:
        c = (a + b) / 2
        if abs(f_prime(c)) < epsilon:
            return c

        if f_prime(c) * f_prime(a) > 0:
            a = c
        else:
            b = c

        iteration += 1

    return (a + b) / 2


print(
    f"Метод половинного деления: x ≈ {bisection_extremum(a, b, epsilon):.{precision}f}"
)

Метод половинного деления: x ≈ 0.5560


## 2. Метод золотого сечения

### Алгоритм:

1. Задаем интервал $[a, b]$, где $a = 0$, $b = \frac{\pi}{4} \approx 0.7854$
2. Вычисляем точки золотого сечения:
   - $x_1 = b - \frac{b-a}{\varphi}$, где $\varphi = \frac{1 + \sqrt{5}}{2} \approx 1.618$
   - $x_2 = a + \frac{b-a}{\varphi}$
3. Вычисляем $f(x_1)$ и $f(x_2)$
4. Если $f(x_1) > f(x_2)$, то $a = x_1$, иначе $b = x_2$
5. Повторяем шаги 2-4

### Вычисления:

**Итерация 1:**

- $a_0 = 0$, $b_0 = 0.7854$
- $x_1 = 0.7854 - \frac{0.7854-0}{1.618} \approx 0.7854 - 0.4854 = 0.3000$
- $x_2 = 0 + \frac{0.7854-0}{1.618} \approx 0.4854$
- $f(x_1) = f(0.3000) = \ln(1 + 0.3000^2) - \sin(0.3000) \approx 0.0870 - 0.2955 \approx -0.2085$
- $f(x_2) = f(0.4854) = \ln(1 + 0.4854^2) - \sin(0.4854) \approx 0.2121 - 0.4646 \approx -0.2525$
- Так как $f(x_1) > f(x_2)$, то $a_1 = x_1 = 0.3000$, $b_1 = b_0 = 0.7854$

**Итерация 2:**

- $a_1 = 0.3000$, $b_1 = 0.7854$
- $x_1 = 0.7854 - \frac{0.7854-0.3000}{1.618} \approx 0.7854 - 0.3000 = 0.4854$
- $x_2 = 0.3000 + \frac{0.7854-0.3000}{1.618} \approx 0.3000 + 0.3000 = 0.6000$
- $f(x_1) = f(0.4854) \approx -0.2525$
- $f(x_2) = f(0.6000) \approx 0.3075 - 0.5646 \approx -0.2571$
- Так как $f(x_1) > f(x_2)$, то $a_2 = x_1 = 0.4854$, $b_2 = b_1 = 0.7854$

**Итерация 3:**

- $a_2 = 0.4854$, $b_2 = 0.7854$
- $x_1 = 0.7854 - \frac{0.7854-0.4854}{1.618} \approx 0.7854 - 0.1854 = 0.6000$
- $x_2 = 0.4854 + \frac{0.7854-0.4854}{1.618} \approx 0.4854 + 0.1854 = 0.6708$
- $f(x_1) = f(0.6000) \approx -0.2571$
- $f(x_2) = f(0.6708) \approx 0.3727 - 0.6234 \approx -0.2507$
- Так как $f(x_1) < f(x_2)$, то $a_3 = a_2 = 0.4854$, $b_3 = x_2 = 0.6708$

**Итерация 4:**

- $a_3 = 0.4854$, $b_3 = 0.6708$
- $x_1 = 0.6708 - \frac{0.6708-0.4854}{1.618} \approx 0.6708 - 0.1146 = 0.5562$
- $x_2 = 0.4854 + \frac{0.6708-0.4854}{1.618} \approx 0.4854 + 0.1146 = 0.6000$
- $f(x_1) = f(0.5562) \approx 0.2723 - 0.5299 \approx -0.2576$
- $f(x_2) = f(0.6000) \approx -0.2571$
- Так как $f(x_1) < f(x_2)$, то $a_4 = a_3 = 0.4854$, $b_4 = x_2 = 0.6000$

**Итерация 5:**

- $a_4 = 0.4854$, $b_4 = 0.6000$
- $x_1 = 0.6000 - \frac{0.6000-0.4854}{1.618} \approx 0.6000 - 0.0708 = 0.5292$
- $x_2 = 0.4854 + \frac{0.6000-0.4854}{1.618} \approx 0.4854 + 0.0708 = 0.5562$
- $f(x_1) = f(0.5292) \approx 0.2504 - 0.5060 \approx -0.2556$
- $f(x_2) = f(0.5562) \approx -0.2576$
- Так как $f(x_1) > f(x_2)$, то $a_5 = x_1 = 0.5292$, $b_5 = b_4 = 0.6000$

После 5 итераций получаем интервал $[0.5292, 0.6000]$, в котором находится минимум. Приближенное значение минимума: $x \approx 0.5646$.


In [82]:
def golden_section_search(a, b, tol, max_iter=1000, find_minimum=True):
    """
    Находит экстремум функции методом золотого сечения.

    Параметры:
    a, b -- границы интервала
    tol -- требуемая точность
    max_iter -- максимальное число итераций
    find_minimum -- True для поиска минимума, False для поиска максимума

    Возвращает:
    x -- приближенное значение экстремума
    """
    # Золотое соотношение
    phi = (1 + 5**0.5) / 2

    # Вычисляем начальные точки золотого сечения
    x1 = b - (b - a) / phi
    x2 = a + (b - a) / phi

    # Вычисляем значения функции
    f1 = f(x1)
    f2 = f(x2)

    iteration = 0

    while abs(b - a) > tol and iteration < max_iter:
        if (f1 > f2) == find_minimum:
            # Минимум находится в правой части
            a = x1
            x1 = x2
            f1 = f2
            x2 = a + (b - a) / phi
            f2 = f(x2)
        else:
            # Минимум находится в левой части
            b = x2
            x2 = x1
            f2 = f1
            x1 = b - (b - a) / phi
            f1 = f(x1)

        iteration += 1

    # Возвращаем середину окончательного интервала
    return (a + b) / 2


print(
    f"Метод золотого сечения: x ≈ {golden_section_search(a, b, epsilon):.{precision}f}"
)

Метод золотого сечения: x ≈ 0.5560


## 3. Метод хорд

### Алгоритм:

1. Выбираем два начальных приближения $x_0$ и $x_1$, для которых $f'(x_0) \cdot f'(x_1) < 0$
2. Вычисляем следующее приближение по формуле:
   $x_{n+1} = x_n - f'(x_n) \cdot \frac{x_n - x_{n-1}}{f'(x_n) - f'(x_{n-1})}$
3. Проверяем условие остановки: $|x_{n+1} - x_n| < \varepsilon$ или $|f'(x_{n+1})| < \varepsilon$
4. Повторяем шаги 2-3

### Вычисления:

**Итерация 1:**

- $x_0 = 0$, $x_1 = 0.7854$
- $f'(x_0) = -1$, $f'(x_1) = 0.2644$
- $x_2 = x_1 - f'(x_1) \cdot \frac{x_1 - x_0}{f'(x_1) - f'(x_0)} = 0.7854 - 0.2644 \cdot \frac{0.7854 - 0}{0.2644 - (-1)} = 0.7854 - 0.2644 \cdot \frac{0.7854}{1.2644} \approx 0.7854 - 0.1642 = 0.6212$
- $f'(x_2) = f'(0.6212) \approx 0.0832 > 0$

**Итерация 2:**

- $x_1 = 0.6212$, $x_2 = 0.7854$ (обновляем значения)
- $f'(x_1) = 0.0832$, $f'(x_2) = 0.2644$
- $x_3 = x_2 - f'(x_2) \cdot \frac{x_2 - x_1}{f'(x_2) - f'(x_1)} = 0.6212 - 0.0832 \cdot \frac{0.6212 - 0.7854}{0.0832 - 0.2644} \approx 0.6212 - 0.0832 \cdot \frac{-0.1642}{-0.1812} \approx 0.6212 - 0.0832 \cdot 0.9061 \approx 0.6212 - 0.0754 = 0.5457$
- $f'(x_3) = f'(0.5457) \approx -0.0138 < 0$

**Итерация 3:**

- $x_2 = 0.5457$, $x_3 = 0.6212$ (обновляем значения)
- $f'(x_2) = -0.0138$, $f'(x_3) = 0.0832$
- $x_4 = x_3 - f'(x_3) \cdot \frac{x_3 - x_2}{f'(x_3) - f'(x_2)} = 0.5457 - (-0.0138) \cdot \frac{0.5457 - 0.6212}{-0.0138 - 0.0832} \approx 0.5457 + 0.0138 \cdot \frac{-0.0755}{-0.0970} \approx 0.5457 + 0.0138 \cdot 0.7784 \approx 0.5457 + 0.0107 = 0.5564$
- $f'(x_4) = f'(0.5564) \approx 0.0006 > 0$

Поскольку $|f'(x_4)| = 0.0006 < 0.001$, мы достигли требуемой точности за 3 итерации.

## Результат метода хорд

Метод хорд дает приближенное значение минимума: $x \approx 0.5564$ с точностью $\varepsilon = 0.001$.


In [83]:
def chord_method_extremum(x0, x1, epsilon, max_iterations=1000):
    """
    Находит экстремум функции методом хорд.

    Параметры:
    x0, x1 -- начальные приближения
    epsilon -- требуемая точность
    max_iterations -- максимальное число итераций

    Возвращает:
    x -- приближенное значение экстремума
    """
    f_prime_x0 = f_prime(x0)
    f_prime_x1 = f_prime(x1)

    # Проверяем, что производная меняет знак на интервале
    if f_prime_x0 * f_prime_x1 >= 0:
        raise ValueError("Производная должна менять знак между x0 и x1")

    iteration = 0
    while (
        abs(x1 - x0) > epsilon
        and abs(f_prime_x1) > epsilon
        and iteration < max_iterations
    ):
        # Вычисляем новое приближение
        x_new = x1 - f_prime_x1 * (x1 - x0) / (f_prime_x1 - f_prime_x0)

        # Обновляем значения
        x0, x1 = x1, x_new
        f_prime_x0, f_prime_x1 = f_prime_x1, f_prime(x1)

        iteration += 1

    return x1


print(f"Метод хорд: x ≈ {chord_method_extremum(0, math.pi / 4, epsilon):.{precision}f}")

Метод хорд: x ≈ 0.5560


## 4. Метод Ньютона

### Алгоритм:

1. Выбираем начальное приближение $x_0$
2. Вычисляем следующее приближение по формуле:
   $x_{n+1} = x_n - \frac{f'(x_n)}{f''(x_n)}$
3. Проверяем условие остановки: $|x_{n+1} - x_n| < \varepsilon$ или $|f'(x_{n+1})| < \varepsilon$
4. Повторяем шаги 2-3

### Вычисления:

Выбираем начальное приближение:

- $x_0 = \frac{\pi}{4} \approx 0.7854$

### Итерации метода Ньютона:

**Итерация 1:**

- $x_0 = 0.7854$
- $f'(x_0) = f'(0.7854) \approx \frac{2 \cdot 0.7854}{1+0.7854^2} - \cos(0.7854) \approx \frac{1.5708}{1.6168} - 0.7071 \approx 0.9716 - 0.7071 \approx 0.2644$
- $f''(x_0) = f''(0.7854) \approx \frac{2-2 \cdot 0.7854^2}{(1+0.7854^2)^2} + \sin(0.7854) \approx \frac{2-1.2333}{(1.6168)^2} + 0.7071 \approx \frac{0.7667}{2.6140} + 0.7071 \approx 0.2932 + 0.7071 \approx 1.0003$
- $x_1 = x_0 - \frac{f'(x_0)}{f''(x_0)} = 0.7854 - \frac{0.2644}{1.0003} \approx 0.7854 - 0.2644 = 0.5211$
- $f'(x_1) = f'(0.5211) \approx -0.0477$

**Итерация 2:**

- $x_1 = 0.5211$
- $f'(x_1) = -0.0477$
- $f''(x_1) = f''(0.5211) \approx \frac{2-2 \cdot 0.5211^2}{(1+0.5211^2)^2} + \sin(0.5211) \approx \frac{2-0.5431}{(1.2715)^2} + 0.4990 \approx \frac{1.4569}{1.6167} + 0.4990 \approx 0.9011 + 0.4990 \approx 1.3990$
- $x_2 = x_1 - \frac{f'(x_1)}{f''(x_1)} = 0.5211 - \frac{-0.0477}{1.3990} \approx 0.5211 + 0.0341 = 0.5552$
- $f'(x_2) = f'(0.5552) \approx -0.0011$

**Итерация 3:**

- $x_2 = 0.5552$
- $f'(x_2) = -0.0011$
- $f''(x_2) = f''(0.5552) \approx \frac{2-2 \cdot 0.5552^2}{(1+0.5552^2)^2} + \sin(0.5552) \approx \frac{2-0.6165}{(1.3082)^2} + 0.5284 \approx \frac{1.3835}{1.7114} + 0.5284 \approx 0.8084 + 0.5284 \approx 1.3356$
- $x_3 = x_2 - \frac{f'(x_2)}{f''(x_2)} = 0.5552 - \frac{-0.0011}{1.3356} \approx 0.5552 + 0.0008 = 0.5560$
- $f'(x_3) = f'(0.5560) \approx 0.0000$

Поскольку $|f'(x_3)| \approx 0 < 0.001$, мы достигли требуемой точности за 3 итерации.

Метод Ньютона дает приближенное значение минимума: $x \approx 0.5560$ с точностью $\varepsilon = 0.001$.


In [84]:
def newton_method_extremum(x0, epsilon, max_iterations=1000):
    """
    Находит экстремум функции методом Ньютона.
    
    Параметры:
    x0 -- начальное приближение
    epsilon -- требуемая точность
    max_iterations -- максимальное число итераций
    
    Возвращает:
    x -- приближенное значение экстремума
    """
    x = x0
    iteration = 0
    
    while iteration < max_iterations:
        # Вычисляем значения производных
        fp = f_prime(x)
        fpp = f_double_prime(x)
        
        # Проверяем, не равна ли вторая производная нулю
        if abs(fpp) < 1e-10:
            raise ValueError("Вторая производная близка к нулю, метод Ньютона не применим")
        
        # Вычисляем новое приближение
        x_new = x - fp / fpp
        
        # Проверяем условие остановки
        if abs(x_new - x) < epsilon or abs(fp) < epsilon:
            return x_new
        
        x = x_new
        iteration += 1
    
    return x


print(f"Метод Ньютона: x ≈ {newton_method_extremum(math.pi/4, epsilon):.{precision}f}")

Метод Ньютона: x ≈ 0.5560


# Вывод по методам поиска экстремума функции

## Сравнительный анализ методов

Мы применили четыре численных метода для нахождения минимума функции $f(x) = \ln(1 + x^2) - \sin(x)$ на интервале $[0, \frac{\pi}{4}]$ с точностью $\varepsilon = 0.001$:

| Метод                     | Результат          | Значение функции            | Значение производной           | Число итераций |
| ------------------------- | ------------------ | --------------------------- | ------------------------------ | -------------- |
| Метод половинного деления | $x \approx 0.5523$ | $f(0.5523) \approx -0.2576$ | $f'(0.5523) \approx -0.0014$   | 5+             |
| Метод золотого сечения    | $x \approx 0.5646$ | $f(0.5646) \approx -0.2585$ | $f'(0.5646) \approx 0.0169$    | 5+             |
| Метод хорд                | $x \approx 0.5560$ | $f(0.5560) \approx -0.2578$ | $f'(0.5560) \approx 0.0000043$ | 4              |
| Метод Ньютона             | $x \approx 0.5560$ | $f(0.5560) \approx -0.2578$ | $f'(0.5560) \approx 0.0000$    | 3              |

## Основные выводы

1. **Точность результатов**: Все методы сходятся к близким значениям минимума функции в окрестности точки $x \approx 0.55$, что подтверждает правильность вычислений. Наиболее точные результаты дают методы хорд и Ньютона, которые сходятся к одному и тому же значению $x \approx 0.5560$.

2. **Скорость сходимости**:

   - Метод Ньютона показал наилучшую скорость сходимости, достигнув требуемой точности за 3 итерации.
   - Метод хорд потребовал 4 итерации.
   - Методы половинного деления и золотого сечения требуют больше итераций (5+) и сходятся медленнее.

3. **Требования к функции**:

   - Метод Ньютона требует вычисления как первой, так и второй производной функции, что может быть сложно для некоторых функций.
   - Метод хорд требует только первой производной.
   - Методы половинного деления и золотого сечения не требуют вычисления производных, что делает их более универсальными.

4. **Чувствительность к начальному приближению**:

   - Метод Ньютона наиболее чувствителен к выбору начального приближения, но в данном случае сработал отлично.
   - Методы половинного деления и золотого сечения требуют только указания интервала, содержащего экстремум.

5. **Точность определения экстремума**:
   - Методы хорд и Ньютона дали значения с очень малым значением производной ($|f'(x)| \approx 0$), что указывает на высокую точность найденного экстремума.
   - Методы половинного деления и золотого сечения дали приближения с бо́льшим значением производной.

## Заключение

Для данной функции $f(x) = \ln(1 + x^2) - \sin(x)$ минимум находится в точке $x \approx 0.5560$ со значением функции $f(0.5560) \approx -0.2578$. Наиболее эффективным методом поиска оказался метод Ньютона, который сошелся к решению за 3 итерации с высокой точностью. Метод хорд также показал хорошие результаты, сойдясь к тому же значению за 4 итерации.
